[SPARK-50793][SQL] Fix MySQL cast function for DOUBLE, LONGTEXT, BIGINT and BLOB types

sunxiaoguang opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Fix MySQL cast function for DOUBLE, LONGTEXT, SMALLINT, INTEGER, BIGINT and BLOB types

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
In MySQL, the type identifier used in cast function is different from their type name. MySQL connector generates SQL which fails with cast.
These are the steps to reproduce.
1. CREATE TABLE test_cast(string_col STRING, long_col LONG, binary_col BINARY, double_col DOUBLE);
2. INSERT INTO test_cast VALUES('0', 0, x'30', 0.0);
3. SELECT * FROM test_cast WHERE CAST(string_col AS BINARY) = binary_col;
4. SELECT * FROM test_cast WHERE CAST(string_col AS LONG) = long_col;
5. SELECT * FROM test_cast WHERE CAST(long_col AS STRING) = '0';
6. SELECT * FROM test_cast WHERE CAST(binary_col AS STRING) = '0';
7. SELECT * FROM test_cast WHERE CAST(double_col AS STRING) = '0';
8. SELECT * FROM test_cast WHERE CAST(string_col AS DOUBLE) = double_col;
9. SELECT * FROM test_cast WHERE CAST(long_col AS DOUBLE) = double_col;

<img width="1236" alt="image" src="https://github.com/user-attachments/assets/d09623f7-947e-4185-b0bb-e0631f54a513" />
<img width="1260" alt="image" src="https://github.com/user-attachments/assets/262c7248-e184-4fe3-b2e5-946554f786f1" />
<img width="1260" alt="image" src="https://github.com/user-attachments/assets/242e8ee6-b62e-4cc8-9ed8-f0456d293593" />
<img width="1260" alt="image" src="https://github.com/user-attachments/assets/a5889f68-d1b5-436f-8c6e-30026a5089f1" />
<img width="1208" alt="image" src="https://github.com/user-attachments/assets/18c2906a-8487-4e02-bfa8-0514f4ea2283" />

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
'No'

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Added new integration test

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
'No'

sunxiaoguang committed:
Fix MySQL cast function for some types

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

sunxiaoguang committed:
Support cast as DOUBLE on MySQL 5.7

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

dongjoon-hyun commented:
cc @yaooqinn

yaooqinn commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,37 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+
+      def testCast(castType: String, sourceCol: String, targetCol: String): Unit = {
+        val sql = s"SELECT * FROM $tableName WHERE CAST($sourceCol AS $castType) = $targetCol"
+        assert(spark.sql(sql).collect().length === 1)
```
let's test the actual value

sunxiaoguang replied:
Sure, let me add that.

sunxiaoguang replied:
The test is validating actual value now, PTAL.

yaooqinn commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:
```diff
@@ -112,6 +112,19 @@ private case class MySQLDialect() extends JdbcDialect with SQLConfHelper with No
+        // MySQL uses CHAR in the cast function for the type LONGTEXT
+        case StringType => "CHAR"
+        // MySQL uses SIGNED INTEGER in the cast function for the types SMALLINT, INTEGER and BIGINT
+        case ShortType | IntegerType | LongType => "SIGNED INTEGER"
```
`CAST(v as SIGNED INTEGER ` always produces a bigint according to the MySQL doc:
> Produces a signed [BIGINT](https://dev.mysql.com/doc/refman/8.4/en/integer-types.html) value.

It's looks an unexpected behavior from the spark side when cast some value to smallint or integer, e.g.

"select cast(1L as int)"

sunxiaoguang replied:
You are right, MySQL doesn't support casting to Short and Integer. But luckily, the wire protocol and JDBC driver implementation works with the same BIGINT result for Short, Integer and Long types.

```java
import java.sql.*;

public class JDBC {
  public static void main(String[] args) throws Exception {
    String host = "localhost";
    int port = 3306;
    String username = "root";
    String password = ""; 
    String database = "test";

    try (Connection conn = DriverManager.getConnection(
          String.format("jdbc:mysql://%s:%d/%s", host, port, database), 
          username, password);
         Statement stmt = conn.createStatement();
         ResultSet rs = stmt.executeQuery("SELECT CAST(1 AS SIGNED INTEGER)")) {
      if (rs.next()) {
        System.out.println("Short: " + rs.getShort(1));
        System.out.println("Integer: " + rs.getInt(1));
        System.out.println("Long: " + rs.getLong(1));
      }   
    }   
  }
}
```
And here is the output
```bash
java -cp mysql-connector-java-8.0.25.jar JDBC.java
Short: 1
Integer: 1
Long: 1
```

yaooqinn replied:
The example above is not sufficient to prove this is OK. Not only is the result set itself used by Spark to fetch results, but its schema is also used for type mapping.

sunxiaoguang replied:
From the new test on actual value and data types, we know the they are in the right types Spark is expecting.

sunxiaoguang replied:
Data types are validated as well in the test to make sure in the case MySQL only cast to BIGINT, we always get the right types defined from Spark side.

```scala
......
      val shortValue = 0.toShort
      val integerValue = 0
      val longValue = 0L
......
      def testCast(castType: String, sourceCol: String, targetCol: String,
                   sourceValue: Any, targetValue: Any): Unit = {
        val sql =
          s"""SELECT $sourceCol, CAST($sourceCol AS $castType) FROM $tableName
             |WHERE CAST($sourceCol AS $castType) = $targetCol""".stripMargin
        val rows = spark.sql(sql).collect()
        assert(rows.length === 1, s"Failed to cast $sourceCol to $castType")
        val row = rows(0)
        assert(row.get(0) === sourceValue, s"$sourceCol does not equal to $sourceValue")
        assert(row.get(0).getClass == sourceValue.getClass,
          s"$sourceCol has different type from $sourceValue")
        assert(row.get(1) === targetValue,
          s"CAST($sourceCol AS $castType) does not equal to $targetValue")
        assert(row.get(1).getClass == targetValue.getClass,
          s"CAST($sourceCol AS $castType) has different type from $targetValue")
      }
......
      testCast("SHORT", "string_col", "short_col", stringValue, shortValue)
      testCast("INTEGER", "string_col", "integer_col", stringValue, integerValue)
      testCast("LONG", "string_col", "long_col", stringValue, longValue)
```

sunxiaoguang replied:
And the test passed locally, let's wait for CI to finish.

<img width="597" alt="image" src="https://github.com/user-attachments/assets/29786f2f-3841-4a5f-9ed7-8c861f222744" />

yaooqinn replied:
You are wrong about what the ResultSetSchema is

sunxiaoguang replied:
I see, there are so many details underlying which I don't know yet. If there are dependencies on the metadata of resultset, it looks like we can't support cast to Short and Integer anymore. How about we remove these two types and leave them not supported at this time, what do you think?

yaooqinn replied:
+1

sunxiaoguang replied:
They are removed from both connector and the integration test, PTAL. Thanks

sunxiaoguang replied:
I changed the title of this PR and the issue to be consistent with the types fixed in this PR.

yaooqinn commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:
```diff
@@ -259,6 +272,8 @@ private case class MySQLDialect() extends JdbcDialect with SQLConfHelper with No
     case ByteType => Option(JdbcType("TINYINT", java.sql.Types.TINYINT))
     case ShortType => Option(JdbcType("SMALLINT", java.sql.Types.SMALLINT))
+    // We override getJDBCType so that DoubleType is mapped to DOUBLE instead.
+    case DoubleType => Option(JdbcType("DOUBLE", java.sql.Types.DOUBLE))
```
Is this change covered in the PR title or desc?

sunxiaoguang replied:
Sorry, I added this later on when testing cast of types on MySQL 5.7. You are right, let me add this type to PR title and the issue description.

sunxiaoguang replied:
I also added comment to this line, so we can understand why this change is necessary.

sunxiaoguang committed:
Make test validating the casted value and type

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

sunxiaoguang committed:
Verify data types in test

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

sunxiaoguang committed:
Do not support cast to short and integer

Do not support cast to short and integer as they break the metadata of
the ResultSet.

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

yaooqinn commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:
```diff
@@ -112,6 +112,21 @@ private case class MySQLDialect() extends JdbcDialect with SQLConfHelper with No
+        case LongType => "SIGNED INTEGER"
+        // MySQL uses BINARY in the cast function for the type BLOB
+        case BinaryType => "BINARY"
+        case _ => getJDBCType(dataType).map(_.databaseTypeDefinition).getOrElse(dataType.typeName)
```
Shall we fail for SMALLINT, INTEGER directly?

sunxiaoguang replied:
Sure, we can add an explicit error message for this case. But I'm not quite familiar with Spark code base right now. I saw there are things like _LEGACY_ERROR_TEMP_xxxx in code base. Could you please leave me some hints on the rules to follow. Thanks.

yaooqinn replied:
If I understand correctly, exceptions here will be captured later and stopping CAST expressions to be pushed down to MySQL.

sunxiaoguang replied:
Sure, I saw other dialects are using both predefined functions in QueryExecutionErrors and classifyException with structured metadata. I'm trying to use QueryExecutionErrors.notSupportTypeError here, not sure if it's appropriate. If not, please let me know and I will try changing it to classifyException instead. PTAL

sunxiaoguang replied:
It happened most of times I can just rerun the test when there are flaky test failed to get checks passed. But there are cases even rerun success, the check still fails. All the tests passed last time and I'm trying to rerun the test again to make the checks pass. It's going to take roughly another two hours.

beliefer replied:
Is cast(SMALLINT, SIGNED INTEGER) or cast(INTEGER, SIGNED INTEGER) legal?

sunxiaoguang replied:
Yes, they are valid statements. The problem comes from different set of type identifier. The source type is not important, for example we can do this for string type as well: 
```sql
SELECT cast('123' AS SIGNED INTEGER)
```

beliefer replied:
If so, I think we can transform SMALLINT and INTEGER to "SIGNED INTEGER" too.

yaooqinn replied:
Aren't SMALLINT and INTEGER target types?

beliefer replied:
Yeah. Got it.

sunxiaoguang committed:
Throw exception with not supported types

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

sunxiaoguang committed:
Check if filter is pushed in test

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

yaooqinn commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:
```diff
@@ -112,6 +112,22 @@ private case class MySQLDialect() extends JdbcDialect with SQLConfHelper with No
+        // To avoid breaking code relying on ResultSet metadata, we support BIGINT only at
+        // this time.
+        case LongType => "SIGNED INTEGER"
+        case ShortType | IntegerType => QueryExecutionErrors.notSupportTypeError(dataType)
```
```suggestion
        case ShortType | IntegerType => QueryExecutionErrors.unsupportedJdbcTypeError(dataType.catalogString)
```

yaooqinn replied:
```suggestion
        case ShortType | IntegerType =>
            QueryExecutionErrors.unsupportedJdbcTypeError(dataType.catalogString)
```

yaooqinn committed:
Update sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala

yaooqinn commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,56 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+             |WHERE CAST($sourceCol AS $castType) = $targetCol""".stripMargin
+        val df = spark.sql(sql)
+        checkFilterPushed(df)
+        val rows = df.collect()
```
Let's use checkAnswer instead of collect

sunxiaoguang replied:
After taking a look at the checkAnswer implementation, it is using `==` to compare any other types than those types need special handling. This means the check may skip actual types, so let's double check if things like this is acceptable.
```scala
 val i = 1
 val s = 1.toShort
 val l = 1L

 println(i == s)
 println(i == l)
 println(s ==l)

The output for these lines of code is:
true
true
true
```


FYI: This is the implementation checkAnswer finally call to compare
```scala
  def compare(obj1: Any, obj2: Any): Boolean = (obj1, obj2) match {
    case (null, null) => true
    case (null, _) => false
    case (_, null) => false
    case (a: Array[_], b: Array[_]) =>
      a.length == b.length && a.zip(b).forall { case (l, r) => compare(l, r)}
    case (a: Map[_, _], b: Map[_, _]) =>
      a.size == b.size && a.keys.forall { aKey =>
        b.keys.find(bKey => compare(aKey, bKey)).exists(bKey => compare(a(aKey), b(bKey)))
      }
    case (a: Iterable[_], b: Iterable[_]) =>
      a.size == b.size && a.zip(b).forall { case (l, r) => compare(l, r)}
    case (a: Product, b: Product) =>
      compare(a.productIterator.toSeq, b.productIterator.toSeq)
    case (a: Row, b: Row) =>
      compare(a.toSeq, b.toSeq)
    // 0.0 == -0.0, turn float/double to bits before comparison, to distinguish 0.0 and -0.0.
    case (a: Double, b: Double) =>
      java.lang.Double.doubleToRawLongBits(a) == java.lang.Double.doubleToRawLongBits(b)
    case (a: Float, b: Float) =>
      java.lang.Float.floatToRawIntBits(a) == java.lang.Float.floatToRawIntBits(b)
    case (a, b) => **a == b**
  }
```

yaooqinn replied:
You can add another check for df.schema for type verification

sunxiaoguang replied:
Got it, I made some changes to use checkAnswer now. But there is another issue with schema, it looks like pretty hard to compare whole schema structure as the Metadata are flexible for different combinations of the test cases. So I just extracted the dataType of StructField in df.schema and checked the data types only.   

PTAL, thanks.

sunxiaoguang committed:
Check result with checkAnswer

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

yaooqinn commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:
```diff
@@ -259,6 +276,10 @@ private case class MySQLDialect() extends JdbcDialect with SQLConfHelper with No
     case StringType => Option(JdbcType("LONGTEXT", java.sql.Types.LONGVARCHAR))
     case ByteType => Option(JdbcType("TINYINT", java.sql.Types.TINYINT))
     case ShortType => Option(JdbcType("SMALLINT", java.sql.Types.SMALLINT))
+    // Because MySQL 5.7 only supports DOUBLE in the CAST function, and DOUBLE is a
```
Is MySQL 5.7 the only exception？

sunxiaoguang replied:
Yes, here are snapshots about how different versions behave.
**MySQL 5.7**
![image](https://github.com/user-attachments/assets/51c0fd05-0969-4550-953c-1740e5e7b29e)
**MySQL 8.0**
![image](https://github.com/user-attachments/assets/1a7e5f64-4f95-4764-9eec-f02d72072b77)
**MySQL 9.1**
<img width="478" alt="image" src="https://github.com/user-attachments/assets/cd97ab59-5241-44f2-9de8-2e5a960e21cb" />

sunxiaoguang replied:
Even though 5.7 is officially end of life, there are still many people using it. Therefore I added support to MySQL 5.7. But I totally understand if we want to get rid of legacy and it's ok to remove this rewrite.

yaooqinn replied:
OK, thank you for the infomation

yaooqinn commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:
```diff
@@ -112,6 +112,23 @@ private case class MySQLDialect() extends JdbcDialect with SQLConfHelper with No
+        // To avoid breaking code relying on ResultSet metadata, we support BIGINT only at
+        // this time.
+        case LongType => "SIGNED INTEGER"
+        case ShortType | IntegerType =>
+            QueryExecutionErrors.unsupportedJdbcTypeError(dataType.catalogString)
```
Can we add a test for not being pushed?

sunxiaoguang replied:
Sure, let me add a test for these cases.

sunxiaoguang replied:
The test for not supported types are covered, PTAL.

beliefer replied:
Personally, LongType can transformed to "SIGNED INTEGER", but ShortType and IntegerType, looks weird.
@yaooqinn How about your opinion?

yaooqinn commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,58 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+      val doubleValue = 0.0
+      val doubleLiteral = "0.0"
+      // CREATE table to use types defined in Spark SQL
+      sql(s"""CREATE TABLE $tableName (
```
nit: let's avoid this multiline string

sunxiaoguang replied:
The template is longer than 100 limit and fails the style check. Unless we change column name to something like c1,c2,c3,c4. What do you think about changing to these column names?

sunxiaoguang replied:
From the next comment I kind of get a idea that you might mean using string concatenation instead of multiline string. Since it works anyway, let me make the changes based on this assumption. In the case of wrong, I would be happy to change these accordingly.

sunxiaoguang replied:
I pushed a commit to concatenate two strings, PTAL

sunxiaoguang commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,58 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+                   sourceDataType: DataType, sourceValue: Any,
+                   targetDataType: DataType, targetValue: Any): Unit = {
+        val sql =
+          s"""SELECT $sourceCol AS source, CAST($sourceCol AS $castType) AS target FROM $tableName
+             |WHERE CAST($sourceCol AS $castType) = $targetCol""".stripMargin
```
Hm, I'm seeing other tests are using string concatenation instead. Just want to make sure I understand it correctly, is that not considered multiline string and is the preferred way to generate long text?

sunxiaoguang replied:
I pushed a commit to concatenate two strings, PTAL

sunxiaoguang committed:
Add test for not supported casts

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

yaooqinn commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,72 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+        val df = spark.sql(sql)
+        castType match {
+          case "SHORT" | "INTEGER" =>
+            checkError(
+              exception = intercept[SparkException] {
+                df.collect()
+              },
+              condition = null
+            )
```
Hmm? Should these cases still be positive without cast-pushdown?

sunxiaoguang replied:
I put cast in projection, so they are triggered no matter if predicate pushdown is enabled. I thought this way we are sure the guard actually works no matter predicate pushdown is enabled or not. In addition, if we don't add the explicit exception, using cast still fails but just with different root cause exception during execution time.

yaooqinn replied:
Why not use valid cases?

sunxiaoguang replied:
Oops, I guess I misunderstood your comment. I thought you want to test the case when unsupported types are used. So did you mean we add test with predicate pushdown disabled?

sunxiaoguang replied:
In case I make mistake again, here is what my understanding is. Let's see which part should be changed.

1. It looks like `visitCast` is always called without configuration item to opt out.
2. We added guards in `visitCast` to throw exceptions explicitly for two types `SHORT` and `INTEGER` to avoid CAST being used in projection and producing incorrect type metadata accidentally.
3. We want to validate the cases where `DOUBLE`, `LONGTEXT`, `BIGINT` and `BLOB` types are used in `cast` and producing the expected resultset including the types in schema.
4. We want to validate the cases where `SHORT` and `INTEGER` which are explicitly disabled are used which generates error during execution.
5. For other cases not involving cast, `SHORT` and `INTEGER` are going to work without any issue.

sunxiaoguang replied:
Hello @yaooqinn, could you please help me understand the issue? Thanks

sunxiaoguang committed:
Use concatenation instead of multiline literal

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

beliefer commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,67 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+        s"INSERT INTO $tableName VALUES($stringLiteral, $longValue, $binaryLiteral, $doubleValue)")
+
+      def testCast(castType: String, sourceCol: String, targetCol: String,
+                   sourceDataType: DataType, sourceValue: Any,
```
indent

sunxiaoguang replied:
Sure, I used dev/scalafmt to format the code this time. But it actually changed too many unrelated code, so I just kept the changes related to this test.

beliefer commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,67 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+        s"INSERT INTO $tableName VALUES($stringLiteral, $longValue, $binaryLiteral, $doubleValue)")
+
+      def testCast(castType: String, sourceCol: String, targetCol: String,
+                   sourceDataType: DataType, sourceValue: Any,
```
ident

sunxiaoguang committed:
Format with dev/scalafmt

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

beliefer commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,91 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+          targetDataType: DataType,
+          targetValue: Any): Unit = {
+        val sql = s"SELECT $sourceCol AS source, CAST($sourceCol AS $castType) AS target " +
+          s"FROM $tableName WHERE CAST($sourceCol AS $castType) = $targetCol"
```
Shall we only test the `CAST($sourceCol AS $castType)`? It seems no need to test `$sourceCol AS source`.

sunxiaoguang replied:
Hello @beliefer could you please help me confirm if we should remove `$sourceCol as source`, thanks.

beliefer replied:
```
val sql = s"SELECT CAST($sourceCol AS $castType) AS target " +
          s"FROM $tableName WHERE CAST($sourceCol AS $castType) = $targetCol"
```
You can simplify the test SQL. No need to check

sunxiaoguang replied:
Sure, they are removed. PTAL, thanks.

beliefer replied:
I have an idea that no matter the collation is, we convert them to the default collation.
`For example, SELECT CAST($sourceCol AS $castType) collate 'ucs_basic'.`

sunxiaoguang committed:
Do not validate source column data and type

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

beliefer commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,61 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+      testCast("INTEGER", "string_col", "long_col", LongType, longValue)
+      testCast("LONG", "string_col", "long_col", LongType, longValue)
+      // We use stringLiteral to make sure both values are using the same collation
+      testCast("STRING", "long_col", stringLiteral, StringType, stringValue)
```
Why not use `string_col` as target column, but string literal ?

beliefer replied:
Why the collation is differences ?

sunxiaoguang replied:
Since the test runs with both `MariaDB connector/j 2.7.12` and `MySQL connector/j 9.1.0` we need this to make the behavior of test consistent. It's a lot of detail. Simply put, we don't want to change the environment and changing the behavior of test setup, so we need to make the new test works with existing test setup while still validating the test cases.

And the following is the detail for everyone's reference. The reason they are incompatible is that for the test setup, `MariaDB connector/j 2.7.12` uses `utf8mb4_unicode_ci` for `collation_connection`, on the other hand `MySQL connector/j 9.1.0` uses `utf8mb4_0900_ai_ci` for `collation_connection`. The test would fail without using `stringLiteral` at here like this.
<img width="1920" alt="image" src="https://github.com/user-attachments/assets/6e4454b5-ffe0-483e-b506-8a5f6201719b" />

[The implementation to decide](https://github.com/mariadb-corporation/mariadb-connector-j/blob/175b47fa386c3a4acac827ae0a00214ef812c297/src/main/java/org/mariadb/jdbc/internal/protocol/AbstractConnectProtocol.java#L1229) `connection_collation` in `MariaDB connector/j 2.7.12` is like this. The serverLanguage is 255 for test setup of `MySQL server 9.1.0`, therefore the collation chosen is 224 which is `utf8mb4_unicode_ci`

```java
  /**  
   * Default collation used for string exchanges with server.
   *
   * @param serverLanguage server default collation
   * @return collation byte
   */
  private byte decideLanguage(int serverLanguage) {
    // return current server utf8mb4 collation
    if (serverLanguage == 45 // utf8mb4_general_ci
        || serverLanguage == 46 // utf8mb4_bin
        || (serverLanguage >= 224 && serverLanguage <= 247)) {
      return (byte) serverLanguage;
    }    

    if (getMajorServerVersion() == 5 && getMinorServerVersion() <= 1) { 
      // 5.1 version doesn't know 4 bytes utf8
      return (byte) 33; // utf8_general_ci
    }    

    // if server language is utf8mb3, use utf8mb4 equivalent collation
    if (serverLanguage == 33) {
      // utf8mb4_general_ci
      return 45;
    }    

    if (serverLanguage == 83) {
      // utf8mb4_bin
      return 46;
    }    

    if (serverLanguage >= 192 && serverLanguage <= 215) {
      // equivalent utf8mb4 collation
      return (byte) (serverLanguage + 32); 
    }    

    return (byte) 224; // UTF8MB4_UNICODE_CI;
  }
```

On the other hand, [the implementation of](https://github.com/mysql/mysql-connector-j/blob/cf2917ea44ae2e43a4514a33771035aa99de73bf/src/main/core-api/java/com/mysql/cj/CharsetMapping.java#L395) `MySQL connector/j 9.1.0` is a lookup table like this. It uses the collation `utf8mb4_0900_ai_ci` which is the same as `collation_server` and `collation_database` with the test setup.
```java
        // 216..222
        collation[223] = new Collation(223, new String[] { "utf8mb3_general_mysql500_ci", "utf8_general_mysql500_ci" }, 0, MYSQL_CHARSET_NAME_utf8mb3);
        collation[224] = new Collation(224, "utf8mb4_unicode_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[225] = new Collation(225, "utf8mb4_icelandic_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[226] = new Collation(226, "utf8mb4_latvian_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[227] = new Collation(227, "utf8mb4_romanian_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[228] = new Collation(228, "utf8mb4_slovenian_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[229] = new Collation(229, "utf8mb4_polish_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[230] = new Collation(230, "utf8mb4_estonian_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[231] = new Collation(231, "utf8mb4_spanish_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[232] = new Collation(232, "utf8mb4_swedish_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[233] = new Collation(233, "utf8mb4_turkish_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[234] = new Collation(234, "utf8mb4_czech_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[235] = new Collation(235, "utf8mb4_danish_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[236] = new Collation(236, "utf8mb4_lithuanian_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[237] = new Collation(237, "utf8mb4_slovak_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[238] = new Collation(238, "utf8mb4_spanish2_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[239] = new Collation(239, "utf8mb4_roman_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[240] = new Collation(240, "utf8mb4_persian_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[241] = new Collation(241, "utf8mb4_esperanto_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[242] = new Collation(242, "utf8mb4_hungarian_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[243] = new Collation(243, "utf8mb4_sinhala_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[244] = new Collation(244, "utf8mb4_german2_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[245] = new Collation(245, "utf8mb4_croatian_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[246] = new Collation(246, "utf8mb4_unicode_520_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[247] = new Collation(247, "utf8mb4_vietnamese_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[248] = new Collation(248, "gb18030_chinese_ci", 1, MYSQL_CHARSET_NAME_gb18030);
        collation[249] = new Collation(249, "gb18030_bin", 0, MYSQL_CHARSET_NAME_gb18030);
        collation[250] = new Collation(250, "gb18030_unicode_520_ci", 0, MYSQL_CHARSET_NAME_gb18030);
        // 251..254
        collation[255] = new Collation(255, "utf8mb4_0900_ai_ci", 1, MYSQL_CHARSET_NAME_utf8mb4);
        collation[256] = new Collation(256, "utf8mb4_de_pb_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[257] = new Collation(257, "utf8mb4_is_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[258] = new Collation(258, "utf8mb4_lv_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[259] = new Collation(259, "utf8mb4_ro_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[260] = new Collation(260, "utf8mb4_sl_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[261] = new Collation(261, "utf8mb4_pl_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[262] = new Collation(262, "utf8mb4_et_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[263] = new Collation(263, "utf8mb4_es_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[264] = new Collation(264, "utf8mb4_sv_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[265] = new Collation(265, "utf8mb4_tr_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[266] = new Collation(266, "utf8mb4_cs_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[267] = new Collation(267, "utf8mb4_da_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[268] = new Collation(268, "utf8mb4_lt_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[269] = new Collation(269, "utf8mb4_sk_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[270] = new Collation(270, "utf8mb4_es_trad_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
        collation[271] = new Collation(271, "utf8mb4_la_0900_ai_ci", 0, MYSQL_CHARSET_NAME_utf8mb4);
```

beliefer replied:
You can adjust the test case at MySQLOverMariaConnectorIntegrationSuite and MySQLIntegrationSuite individually

sunxiaoguang replied:
The logic here is reasonable from test perspective. If we put them into separate test, there will be duplication. Is that ok?

beliefer replied:
Please write these tests, then I will review again.

sunxiaoguang replied:
It's updated to test cast as string differently on MariaDB, PTAL

sunxiaoguang committed:
Test cast string differently on MariaDB

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

sunxiaoguang committed:
Add comment to test

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

beliefer commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,84 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+        s"CREATE TABLE $tableName ($stringCol STRING, $longCol LONG, " +
+          s"$binaryCol BINARY, $doubleCol DOUBLE)")
+      sql(
+        s"INSERT INTO $tableName VALUES($stringLiteral, $longValue, $binaryLiteral, $doubleValue)")
```
Is there a way insert different string literal into table, and then the query only related to string_col ?
So we can remove the `testCastStringTarget`.

sunxiaoguang replied:
No, it is not possible. 

As Spark doesn't specify collation for tables and columns, the collation of string columns is always the one defined in `collation_database`. On the other hand, the collation of string resulted from cast is the same as `collation_connection`.

`MariaDB Connector/J` uses different `collation_connection` `utf8mb4_unicode_ci ` whereas the `MySQL server 9.1.0` uses `utf8mb4_0900_ai_ci ` for `collation_database`. So we need to use string literal which also uses collation defined in `collation_connection` to make sure we are comparing two strings in the same collation.

For the case `MySQL Connector/J` is used, it uses the same collation for connection as collation for database on the server side, therefore we can use the column directly.

Prior to this commit, we used string literal which behave the same on both MySQL and MariaDB connectors. With this commit, we are doing it differently to compare with string literal for `MariaDB Connector/J` and compare with string column for `MySQL Connector/J`.

beliefer reviewed (commented):
Overall, it looks okay.
@yaooqinn Do you have any other comments?

sunxiaoguang commented:
Hello @yaooqinn 
I'm wondering if you could take some time to look at the latest code and share any comments you might have?

sunxiaoguang commented:
> Thank you @sunxiaoguang, the changes almost look good to me, except for two minor comments.

Hello @beliefer, @yaooqinn might be busy on something and it's been quite a while since he left this comment. If the PR looks good to you, I guess the overall changes look good to him as well. 

How about we merge this PR so we can start working on the pushdown PR #49335?

beliefer commented:
@sunxiaoguang
I have an idea split the support for StringType into another PR. So this PR is good to review and merge.

beliefer commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -264,4 +342,10 @@ class MySQLOverMariaConnectorIntegrationSuite extends MySQLIntegrationSuite {
+  // MariaDB Connector/J uses collation 'utf8mb4_unicode_ci' as collation for connection.
+  // The MySQL server 9.1.0 uses collation 'utf8mb4_0900_ai_ci' for database by default.
+  // This method uses string literal so the result of cast and literal have the same collation.
+  override def testCastStringTarget(stringLiteral: String, stringCol: String): String =
```
I'm not sure if this function is good enough to test cast string.

sunxiaoguang replied:
The root cause is the MariaDB JDBC driver is not able to use the same default collation for connection. If we can accept changing the collation of connection to make it consistent with MySQL server, the test logic can be the same for both MySQL JDBC and MariaDB JDBC.

I didn't want to change the default settings for all tests, but that's probably the only way to fix the inconsistency and use the same code in test. What do you think about this?

sunxiaoguang replied:
Or in other words, it is due to compatibility issue between MariaDB JDBC and MySQL Server 9.

beliefer commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:
```diff
@@ -241,6 +241,84 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTest
+          targetValue: Any): Unit = {
+        val sql = s"SELECT CAST($sourceCol AS $castType) AS target " +
+          s"FROM $tableName WHERE CAST($sourceCol AS $castType) = $targetCol"
+        val df = spark.sql(sql)
```
I have an idea that no matter the collation is, we convert them to the default collation.
`df.select(collate($"target", "ucs_basic"))`

sunxiaoguang replied:
Cool, didn't know Spark SQL has collate function in master branch. I think that suppose to work, let me have a try.

sunxiaoguang replied:
Unfortunately, the `collate` function is not pushed down to JDBC.

sunxiaoguang replied:
Looks like we have no choice but explicitly specifying `collation_connection` on the client side to make it consistent with `collation_database` and `collation_server` on the server side.

After changing the collation of connection, test can be written in the same way with both MySQL Connector/J and MariaDB Connector/J. PTAL, Thanks.

beliefer replied:
I mean collected the result of the query pushed down. And then use Spark to calculate the collate function.
The df is the collected result transformed to data set.
`df.select(collate($"target", "ucs_basic"))`

sunxiaoguang replied:
Due to the mismatch of collation from JDBC client and the MySQL server, the SQL statements, which do not have any collation specified, won't run on MySQL server. So we cannot get result, and cannot do anything on Spark side to process the result.

beliefer replied:
```
val sql = s"SELECT CAST($sourceCol AS $castType) AS target " +
          s"FROM $tableName WHERE CAST($sourceCol AS $castType) = $targetCol"
val df1 = spark.sql(sql)
val data = df.collect()
val df2 = ... // convert dato to dataset
df2.select(collate($"target", "ucs_basic"))
```
We can't do this ?

sunxiaoguang replied:
Unfortunately, the SQL sent to MySQL fails with this error message. Therefore we have no chance to get the result.

<img width="1920" alt="image" src="https://github.com/user-attachments/assets/6e4454b5-ffe0-483e-b506-8a5f6201719b" />

sunxiaoguang replied:
This expression is where MySQL server complains.
`CAST($sourceCol AS $castType) = $targetCol`

The collation on the left hand side is the same as `collation_connection`
The collation on the right hand side it the same as `collation_database`

Since MariaDB Connector/J uses different `collation_connection` than the `collation_database` on MySQL Server. This expression is rejected by MySQL Server.

sunxiaoguang replied:
Hello @beliefer, do you have any comment on this?

beliefer replied:
`CAST($sourceCol AS $castType) = $targetCol`
It seems we can replace the $targetCol with string with different collation.
We can generate the string literal with different collation for test.

sunxiaoguang replied:
Yes, that's what we did for MariaDB to make it work at the beginning. But we would not be able to verify if the cast result matches the expected type stored in target column. Later changes to collation_connection does not give up this check.

sunxiaoguang replied:
Looks like the implementation of collation in Spark is not handled in underlying connectors. Therefore we can not pass collation from Spark to generate different SQL under this circumstance.

beliefer replied:
I think we should create the test table with the MySQL DDL syntax. For example,
```
CREATE TABLE t
(
    c1 VARCHAR(20) CHARACTER SET utf8mb4,
    c2 TEXT CHARACTER SET latin1 COLLATE latin1_general_cs
);
```
And then let `MySQLDialect.getCatalystType` supports convert JDBC collation to Spark collation and `MySQLDialect.getJDBCType` supports convert Spark collation to JDBC collation.
At last we test with
`select * from t where c2 = 'abc'` in Spark. In fact, we will push down the query
 `select * from t where c2 = latin1'abc' COLLATE latin1_general_cs` to MySQL.

beliefer replied:
I think we should create the test table with the MySQL DDL syntax. For example,
```
CREATE TABLE t
(
    c1 VARCHAR(20) CHARACTER SET utf8mb4,
    c2 TEXT CHARACTER SET latin1 COLLATE latin1_general_cs
);
```
And then let `MySQLDialect.getCatalystType` supports convert JDBC collation to Spark collation and `MySQLDialect.getJDBCType` supports convert Spark collation to JDBC collation.
At last we test with
`select * from t where c2 = 'abc'` in Spark. In fact, we will push down the query
 `select * from t where c2 = latin1'abc' COLLATE latin1_general_cs` to MySQL.

beliefer replied:
I want create a PR to give a basic support.

beliefer replied:
You just need supports pushdown the collation to H2 dialect as an example or select MySQL.
Other dialects remains the followup pr.
First, make the DS V2 pushdown framework supports collation and select MySQL or H2 as a basic implementation.
Then continue proceeding this PR.

sunxiaoguang replied:
Hello @beliefer, could you please help me understand what to do the next? 

Collation is tightly coupled with character sets and the combinations to support from MySQL are a lot. In addition, collation support in Spark was recently added with four options only which cannot support all the collations in MySQL. In the case when we want to support collation, we need to have a careful design top down from Spark kernel I guess.

Looking forward to hearing your comments, Thanks.

sunxiaoguang replied:
> You just need supports pushdown the collation to H2 dialect as an example or select MySQL. Other dialects remains the followup pr. First, make the DS V2 pushdown framework supports collation and select MySQL or H2 as a basic implementation. Then continue proceeding this PR.

Ok, I have some quick questions about this. 
- I assume features like this should go through a RFC procedure. I will try to figure it out myself. Meanwhile, I would really appreciate if you can give me some hints of previous work similar as references
- The subtle differences of different collations and encodings can be very tricky, what's your suggestion for tables with collations that we know that Spark don't support yet. How do we work with collations that are a little different but overall behave the same. As a related question, different version of MySQL server support different set of collations. Shell we support the latest MySQL server only or we need to consider the version of MySQL server and try use the collation in that MySQL version.

Really appreciate your help, thanks.

beliefer replied:
> In addition, collation support in Spark was recently added with four options only which cannot support all the collations in MySQL. In the case when we want to support collation, we need to have a careful design top down from Spark kernel I guess.

Sure. Spark can pushdown the collation that both supported by Spark and MySQL. Otherwise, do not pushdown them.

beliefer replied:
The answer for the second question is the same as the one mentioned above.

sunxiaoguang replied:
Let me double check in case I misunderstood. Since people can read data from JDBC Catalog, What do we do for the tables with collations that are not supported in Spark? Shell we stop people from reading data from those tables?

sunxiaoguang replied:
Sure, so we do it in best efforts and give it up when it is not possible.

sunxiaoguang committed:
Update MariaDB Connector/J parameters

Signed-off-by: Xiaoguang Sun <sunxiaoguang@gmail.com>

beliefer commented:
>I assume features like this should go through a RFC procedure. I will try to figure it out myself. Meanwhile, I would really appreciate if you can give me some hints of previous work similar as references

Your can reference the pushdown of Cast, which need compile Spark's data type to underlying data source's one.

sunxiaoguang commented:
> > I assume features like this should go through a RFC procedure. I will try to figure it out myself. Meanwhile, I would really appreciate if you can give me some hints of previous work similar as references
> 
> Your can reference the pushdown of Cast, which need compile Spark's data type to underlying data source's one.

Thanks for the hint. Another question, do we need to write a complete design document and going through a review process about the design before actually writing the implementation?

beliefer commented:
> Thanks for the hint. Another question, do we need to write a complete design document and going through a review process about the design before actually writing the implementation?

I think it is not complicate enough, you just need a try and commit a PR.
